drivers/clock_control/nrf_power_clock.c

```diff
@@ -1,280 +1,327 @@
+	}
 
-		irq_disable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
+	return CLOCK_CONTROL_STATUS_OFF;
+}
 
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+static int clock_stop(struct device *dev, clock_control_subsys_t sub_system)
+{
+	const struct nrf_clock_control_config *config =
+						dev->config->config_info;
+	struct nrf_clock_control *data = dev->driver_data;
+	int err = 0;
+	int key;
+
+	key = irq_lock();
+	data->ref--;
+	if (data->ref == 0) {
+		bool do_stop;
+
+		DBG(dev, "Stopping");
+		sys_slist_init(&data->list);
+
+		do_stop =  (config->stop_handler) ?
+				config->stop_handler(dev) : true;
+
+		if (do_stop) {
+			nrf_clock_task_trigger(config->stop_tsk);
+			nrf_clock_event_clear(config->started_evt);
+		}
 
-		intenset = NRF_CLOCK->INTENSET;
-		nrf_clock_int_enable(NRF_CLOCK_INT_HF_STARTED_MASK);
+		data->started = false;
+	} else if (data->ref < 0) {
+		data->ref = 0;
+		err = -EALREADY;
+	}
 
-		nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTART);
+	irq_unlock(key);
 
-		while (NRF_CLOCK->EVENTS_HFCLKSTARTED == 0) {
-			__WFE();
-			__SEV();
-			__WFE();
-		}
+	return err;
+}
 
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+static bool is_in_list(sys_slist_t *list, sys_snode_t *node)
+{
+	sys_snode_t *item = sys_slist_peek_head(list);
 
-		if (!(intenset & CLOCK_INTENSET_HFCLKSTARTED_Msk)) {
-			nrf_clock_int_disable(NRF_CLOCK_INT_HF_STARTED_MASK);
+	do {
+		if (item == node) {
+			return true;
+		} else {
+			item = sys_slist_peek_next(item);
 		}
+	} while (item);
 
-		NVIC_ClearPendingIRQ(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-
-		irq_enable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-	} else {
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+	return false;
+}
 
-		nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTART);
-	}
+static int clock_async_start(struct device *dev,
+			     clock_control_subsys_t sub_system,
+			     struct clock_control_async_data *data)
+{
+	const struct nrf_clock_control_config *config =
+						dev->config->config_info;
+	struct nrf_clock_control *clk_data = dev->driver_data;
+	int key;
 
-	/* release resource guard */
-	m16src_grd = 0U;
+	__ASSERT_NO_MSG((data == NULL) ||
+			((data != NULL) && (data->cb != NULL)));
 
-hf_already_started:
-	/* rollover should not happen as start and stop shall be
-	 * called in pairs.
-	 */
-	__ASSERT_NO_MSG(m16src_ref);
+	key = irq_lock();
+	clk_data->ref++;
 
-	stat = NRF_CLOCK_HFCLK_HIGH_ACCURACY | CLOCK_HFCLKSTAT_STATE_Msk;
-	if ((NRF_CLOCK->HFCLKSTAT & stat) == stat) {
-		return 0;
+	if (clk_data->started) {
+		irq_unlock(key);
+		if (data) {
+			data->cb(dev, data->user_data);
+		}
 	} else {
-		return -EINPROGRESS;
-	}
-}
+		if (clk_data->ref == 1) {
+			bool do_start;
 
-static int m16src_stop(struct device *dev, clock_control_subsys_t sub_system)
-{
-	u32_t imask;
+			do_start =  (config->start_handler) ?
+					config->start_handler(dev) : true;
 
-	ARG_UNUSED(sub_system);
+			if (do_start) {
+				nrf_clock_event_clear(config->started_evt);
+				nrf_clock_task_trigger(config->start_tsk);
```

Comment:
Where will this task be triggered in case `start_handler` returns `false`?

Reply:
if start_handler returns false then it means that starting clock will be handled by module that provided handler. In that particular case start handler will returns false when it didn't yet stop the clock because calibration was ongoing. It will not call start because lfclk was not yet stopped (`CAL_ACTIVE_OFF` state).